vochain: order transactions when creating new blocks

p4u opened the pull request:
The main objective of the PR was to add transaction ordering to the block production; however, it ended up being a bit more complex, since I found several things that required attention.

1. Transactions are now ordered by sender address and nonce during the Propose block phase. Those transactions which are discarded (because they are invalid) are sent back to the mempool and will be proposed in a further block.

2. The Propose Block method is now checking transactions against a mutable, ongoing state. This means that to consider a list of transactions valid, we need to construct a temporary State which is modified with the transaction's execution. Once the proposal block finishes, it calls State.Rollback() in order to remove the changes.

3. The Process Proposal method checks the validity of transactions and discards the block if some is invalid. This assures we are not including invalid transactions into blocks.

4. There is a new internal TTL mechanism for mempool transactions. After a certain number of blocks, if the transaction did not succeed, it is removed. 

5. The Update SIK roots process now only happens once per block. This is a cleaner approach and prevent consensus issues.

6. The tokens test has been modified to send transactions with reversed nonce, this way we are also testing the ordering of the proposal block creation.

p4u committed:
vochain: order transactions when creating new blocks

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: update sik roots only once per block

Signed-off-by: p4u <pau@dabax.net>

jordipainan commented on vochain/app.go:
```diff
@@ -33,7 +35,10 @@ import (
+	recheckTxHeightInterval = 6 * 5 // 5 minutes
+	// transactionBlocksTTL is the number of blocks after which a transaction is
+	// removed from the mempool.
+	transactionBlocksTTL = 6 * 10 // 10 minutes
```
Are these values related to the consesus block time? If so, I would like to suggest to calculate the values on runtime and depending on the block time related parameters.

p4u replied:
I'm not sure about this approach. Because the same way someone defines a shorter block period to have a faster blockchain, the TTL, Recheck and so on should also happen quickly. So I prefer to base everything on the number of blocks and not on the time between blocks.

mariajdab commented on cmd/end2endtest/account.go:
```diff
@@ -181,23 +184,48 @@ func testSendTokens(api *apiclient.HTTPclient, aliceKeys, bobKeys *ethereum.Sign
+		_, err = alice.TransferWithNonce(state.BurnAddress, 1, aliceAcc.Nonce+1)
+		if err != nil {
+			log.Fatalf("cannot burn tokens: %v", err)
+		}
```
```
if err := alice.TransferWithNonce(state.BurnAddress, 1, aliceAcc.Nonce+1); err != nil {
}
```

Looks good the changes on end2end, maybe avoid use `log.Fatal`, but I can take care of it :raised_hands:

jordipainan commented on vochain/app.go:
```diff
@@ -74,6 +81,9 @@ type BaseApplication struct {
 	dataDir             string
 	genesisInfo         *tmtypes.GenesisDoc
 
+	// prepareProposalLock is used to avoid concurrent calls between PrepareProposal and FinalizeBlock
```
Given that is also used in ProcessProposal, is it right to understand that we try to avoid concurrency issues between PrepareProposal, ProcessProposal and FinalizeBlock ?

p4u replied:
yes

p4u committed:
fix tests regading the SIK changes

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
add missing RegisterSIK tx on extract nonce

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
log: if output is a json file, format propperly

p4u committed:
vochain: set isSync to true while bootstraping the blockchain

Signed-off-by: p4u <pau@dabax.net>

altergui commented on vochain/transaction/nonce.go:
```diff
@@ -0,0 +1,112 @@
+			err = fmt.Errorf("set process payload is nil")
+			break
+		}
+		txNonce = txWithNonce(ptx)
```
i see a lot of deduplication possible in this switch block, what about:
```go
func (t *TransactionHandler) ExtractNonceAndSender(vtx *vochaintx.Tx) (*common.Address, uint32, error) {
	var ptx interface {
		GetNonce() uint32
	}

	switch payload := vtx.Tx.Payload.(type) {
	case *models.Tx_NewProcess:
		ptx = payload.NewProcess
	case *models.Tx_SetProcess:
		ptx = payload.SetProcess
	case *models.Tx_SendTokens:
		ptx = payload.SendTokens
	case *models.Tx_SetAccount:
		if payload.SetAccount.Txtype == models.TxType_CREATE_ACCOUNT {
			// create account tx is a special case where the nonce is not relevant
			return nil, 0, nil
		}
		ptx = payload.SetAccount
	case *models.Tx_CollectFaucet:
		ptx = payload.CollectFaucet
	case *models.Tx_Vote, *models.Tx_Admin, *models.Tx_MintTokens, *models.Tx_SetKeykeeper,
		*models.Tx_SetTransactionCosts, *models.Tx_DelSIK, *models.Tx_RegisterKey, *models.Tx_SetSIK,
		*models.Tx_RegisterSIK:
		// these tx does not have incremental nonce
		return nil, 0, nil
	default:
		// force panic to detect new tx types
		panic(fmt.Sprintf("unknown payload type on extract nonce: %T", payload))
	}

	if ptx == nil {
		return nil, 0, fmt.Errorf("payload is nil")
	}

	pubKey, err := ethereum.PubKeyFromSignature(vtx.SignedBody, vtx.Signature)
	if err != nil {
		return nil, 0, fmt.Errorf("cannot extract public key from vtx.Signature: %w", err)
	}
	addr, err := ethereum.AddrFromPublicKey(pubKey)
	if err != nil {
		return nil, 0, fmt.Errorf("cannot extract address from public key: %w", err)
	}

	return &addr, ptx.GetNonce(), nil
}

```

altergui commented on vochain/transaction/nonce.go:
```diff
@@ -0,0 +1,112 @@
+		return nil, 0, nil
+	default:
+		// force panic to detect new tx types
+		panic(fmt.Sprintf("unknown payload type on extract nonce: %T", payload))
```
is this a debugging leftover?
then maybe just `log.Warnf`

p4u replied:
It's in purpose (as the comment says) to detect new tx types which have not been added here. So it actas as a protection for future new txs.

altergui replied:
but that means if someone broadcasts a transaction of a new type, the whole network is immediately brought down with a panic?

p4u replied:
That's a good point. But I believe Protobuf won't allow the unmarshal or would set it to default value which is models.Tx_UNKNOWN. It is something to check, for sure.

If the protobuf models contain the txType because it is a new introduced tx type, then the tests should fail.

altergui replied:
if protobuf "protects" us from a DDOS, that's very nice, but in any case, we should really avoid panics in the codepath that processes input from outside, i bet it would raise questions from any security auditor

p4u replied:
I have moved the Panic to log.Error(), thanks for the input @altergui

altergui commented on vochain/app.go:
```diff
@@ -506,8 +579,42 @@ func (app *BaseApplication) PrepareProposal(ctx context.Context,
 // Application SHOULD accept a prepared proposal passed via ProcessProposal, even if a part of the proposal
 // is invalid (e.g., an invalid transaction); the Application can ignore the invalid part of the prepared
 // proposal at block execution time. The logic in ProcessProposal MUST be deterministic.
```
this comment says the opposite of what happens inside the func (returns ResponseProcessProposal_REJECT in case a transaction was invalid)

i trust your code, but then the comment needs updating (or an explanation why are you deviating from this SHOULD)

p4u replied:
I guess this comment is for the safety of the non-experneiced users. 

Here more details: https://github.com/cometbft/cometbft/blob/main/spec/abci/abci++_methods.md#processproposal

> Moreover, application implementors SHOULD always set ResponseProcessProposal.status to ACCEPT, unless they really know what the potential liveness implications of returning REJECT are.

So, if we accept any block proposed by a validator, we might run into the risk of including invalid transactions to blocks, which won't be executed since deliverTx() will fail, but the invalid tx will remain as part of the blockchain forever. Thus, a misbehaving validator could potentially fill the blockchain with garbage.

In the other side, if we reject the proposal, we force the validators to act honestly. The downside is that a block will take longer to be produced since a new validator needs to start over the transaction set for a new block.

I think, the way we do it should be safe since on prepare proposal, we are executing all tx against an ongoing state (reverted with Rollback() at the end) and on process proposal we do exactly the same. If the state is deterministic (which must be) the outcome should always be the same (either ACCEPT or REJECT).

altergui commented on vochain/app.go:
```diff
@@ -257,11 +268,23 @@ func (app *BaseApplication) InitChain(_ context.Context,
+	// store the initial height of the tx
+	initialTTLheight, isReferenced := app.txTTLReferences.LoadOrStore(txReference, app.Height())
+	// check if the tx is referenced by a previous block and the TTL has expired
+	if isReferenced && app.Height() > initialTTLheight.(uint32)+transactionBlocksTTL {
```
the `isReferenced` here is redundant, since `app.Height() > initialTTLheight.(uint32)+transactionBlocksTTL` will never be true unless the value was loaded (not stored)
you can simply do

```go
	// retrieve the height when the tx was first checked, or else store the current height
	initialTTLheight, _ := app.txTTLReferences.LoadOrStore(txReference, app.Height())
	// if the tx was checked in a previous block and the TTL has expired
	if app.Height() > initialTTLheight.(uint32)+transactionBlocksTTL {
		// remove tx reference and return checkTx error
```

altergui reviewed (approved):
LGTM,
the only real issue i saw is the `panic` on unknown payloads,

the rest are comments on code readability, that doesn't change behaviour at all

altergui commented on vochain/app.go:
```diff
@@ -474,25 +497,75 @@ func (*BaseApplication) Query(_ context.Context,
 // the ResponsePrepareProposal call. The logic modifying the raw proposal MAY be non-deterministic.
 func (app *BaseApplication) PrepareProposal(ctx context.Context,
```
deepsource says `ctx` is now unused, suggests to name it `_` then

altergui commented on vochain/app.go:
```diff
@@ -257,11 +268,23 @@ func (app *BaseApplication) InitChain(_ context.Context,
+		log.Debugw("pruning expired tx from mempool", "height", app.Height(), "hash", fmt.Sprintf("%x", txReference))
+		app.txTTLReferences.Delete(txReference)
+		return &abcitypes.ResponseCheckTx{Code: 1, Data: []byte(fmt.Sprintf("tx expired %x", txReference))}, nil
+	}
```
also, these lines are not covered by tests: https://coveralls.io/builds/62999336/source?filename=vochain%2Fapp.go#L276

it would be great if you could add a test or extend one of the existing ones. 

AFAIU, pushing a tx with a very high nonce would do it, right?
and then wait until it's pruned, then check somehow the result

i can also self-assign this little task as well if you prefer and push a commit to PR

p4u replied:
please, feel free to add the missing test :)

p4u committed:
arbo: add a cache for the current Root and coupe of locks more

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
end2endtest: tokentxs better check nonce ordering

p4u committed:
some fixes on consensus stability

1. UpdateSIKRoots must be executed on the current State transaction,
so when the commit is done for te block, it is also stored.

2. Prepare/Process proposal check (with panic) if there are pending
transactions to commit.

3. NoState transaction is correctly discarted now when Rollback

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
bump new dev genesis

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
apply PR suggestions

p4u committed:
arbo: remove the root cache

p4u committed:
make vet happy

p4u committed:
vochain: reset txCounter on commit

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
arbo: move some locks to read-locks

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: save state on FinalizeBlock

p4u committed:
replace Commit by CommitState. Add some extra info on shutdown

p4u committed:
vochain: rollback to save state on cometBFT Commit()

p4u committed:
working on consensus failure

p4u committed:
more

p4u committed:
arbo: add a new test for testing root updates on tx

p4u committed:
fix consensus failure

When we migrated to cometBFT 0.37, the new API requires to do the
block execution process in two steps: 1) execute 2) save

Our statedb was not designed to work in two steps, but one.
So there was a bug on the root hash returned by the execute step,
which was different from the one returned by save. That created
a consensus failure if the node was stop when processing transactions.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
bump dev genesis

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: implement optimistic execution

During block production, the validator nodes execute the
block transaction during the proposal phase. This way we don't
need to perform a Rollback() and re-execute again on FinalizeBlock

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
use the cometBFT block-hash for safety execution on FinalizeBlock

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
rollback state when starting PrepareProposal

If a consensus round fails, then PrepareProposal might be
executed for the same height after ProcessProposal. Then we
need to be sure that PrepareProposal starts always from a fresh
state.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
fix prepare proposal test

Signed-off-by: p4u <pau@dabax.net>

mvdan commented on tree/arbo/tree.go:
```diff
@@ -620,6 +625,8 @@ func (t *Tree) SetRootWithTx(wTx db.WriteTx, root []byte) error {
 // The list of roots for a level can be obtained using tree.RootsFromLevel().
 func (t *Tree) Snapshot(fromRoot []byte) (*Tree, error) {
+	t.Lock()
+	defer t.Unlock()
```
I assume these were added while you were debugging the pebble panic, and should now be removed?

mvdan replied:
not sure what happened here - I still think these added locks/unlocks were only there while you debugged pebble

p4u replied:
Yes, you are right. But while I was reading the Arbo code, I found some missing locks. Which might not be critical, but I prefer to make it secure.

mvdan commented on vochain/cometbft.go:
```diff
@@ -0,0 +1,452 @@
+	}
+
+	// Sort the transactions based on the sender's address and nonce
+	sort.Slice(validTxInfos, func(i, j int) bool {
```
FYI now that we depend on Go 1.21 you can use better generics, e.g. https://pkg.go.dev/slices#SortFunc

p4u replied:
I don't feel ready yet to use Go Generics in a comfortable way. I think I'll let this improvement for the future, or feel free if you want to take it.

mvdan commented on vochain/cometbft.go:
```diff
@@ -0,0 +1,452 @@
+			if validTxInfos[i].Addr.String() == validTxInfos[j].Addr.String() {
+				return validTxInfos[i].Nonce < validTxInfos[j].Nonce
+			}
+			return validTxInfos[i].Addr.String() < validTxInfos[j].Addr.String()
```
nit: reuse the calls to String, since they might not be free

mvdan replied:
or even better, can we compare the addresses without turning them to strings? presumably the addresses are bytes, so we can use `bytes.Compare` without needing to turn them to e.g. hex or base64 strings.

mvdan commented on tree/arbo/tree.go:
```diff
@@ -84,7 +84,7 @@ var (
 // Tree defines the struct that implements the MerkleTree functionalities
 type Tree struct {
-	sync.Mutex
+	sync.RWMutex
```
remember that RWMutex is more complex and expensive than Mutex, it is only worth it if we're going to be doing mostly reads. if there is a roughly equal mix of reads and writes, a regular Mutex will perform much better.

p4u replied:
Right, going back to Mutex. It was part of my previous testing.

p4u committed:
apply PR suggestions

p4u committed:
Merge branch 'main' of github.com:vocdoni/vocdoni-node into f/order_transactions

p4u committed:
remove tx from mempool

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove invalid transactions from mempool after three invalid executions
